File: python/pyautd3/gain/holo/constraint.py
"""
File: constraint.py
Project: holo
Created Date: 25/10/2022
Author: Shun Suzuki
-----
Last Modified: 29/09/2023
Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
-----
Copyright (c) 2022-2023 Shun Suzuki. All rights reserved.

"""


from pyautd3.emit_intensity import EmitIntensity
from pyautd3.native_methods.autd3capi_gain_holo import EmissionConstraintPtr
from pyautd3.native_methods.autd3capi_gain_holo import NativeMethods as GainHolo


class EmissionConstraint:
    """Amplitude constraint."""

    _ptr: EmissionConstraintPtr

    def __init__(self: "EmissionConstraint", ptr: EmissionConstraintPtr) -> None:
        self._ptr = ptr

    @staticmethod
    def dont_care() -> "EmissionConstraint":
        """Do nothing (this is equivalent to `clamp(0, 1)`)."""
        return EmissionConstraint(GainHolo().gain_holo_constraint_dot_care())

    @staticmethod
    def normalize() -> "EmissionConstraint":
        """Normalize the value by dividing the maximum value."""
        return EmissionConstraint(GainHolo().gain_holo_constraint_normalize())

    @staticmethod
    def uniform(value: int | EmitIntensity) -> "EmissionConstraint":
        """Set all amplitudes to the specified value.

        Arguments:
        ---------
            value: emission intensity
        """
        return EmissionConstraint(GainHolo().gain_holo_constraint_uniform(EmitIntensity._cast(value).value))

    @staticmethod
    def clamp(min_v: int | EmitIntensity, max_v: int | EmitIntensity) -> "EmissionConstraint":
        """Clamp all amplitudes to the specified range.

        Arguments:
        ---------
            min_v: minimum emission intensity
            max_v: maximum emission intensity
        """
        return EmissionConstraint(
            GainHolo().gain_holo_constraint_clamp(
                EmitIntensity._cast(min_v).value,
                EmitIntensity._cast(max_v).value,
            ),
        )

    def _constraint_ptr(self: "EmissionConstraint") -> EmissionConstraintPtr:
        return self._ptr

File: dotnet/cs/src/Gain/Holo/Constraint.cs
/*
 * File: Constraint.cs
 * Project: Holo
 * Created Date: 13/09/2023
 * Author: Shun Suzuki
 * -----
 * Last Modified: 29/11/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2023 Shun Suzuki. All rights reserved.
 * 
 */

#if UNITY_2018_3_OR_NEWER
#define USE_SINGLE
#endif

using AUTD3Sharp.NativeMethods;

#if USE_SINGLE
using float_t = System.Single;
#else
using float_t = System.Double;
#endif

namespace AUTD3Sharp.Gain.Holo
{
    /// <summary>
    /// Amplitude constraint
    /// </summary>
    public interface IAmplitudeConstraint
    {
        internal EmissionConstraintPtr Ptr();
    }

    /// <summary>
    /// Do nothing (this is equivalent to `Clamp(0, 1)`)
    /// </summary>
    public sealed class DontCare : IAmplitudeConstraint
    {
        EmissionConstraintPtr IAmplitudeConstraint.Ptr() => NativeMethodsGainHolo.AUTDGainHoloConstraintDotCare();
    }

    /// <summary>
    /// Normalize the value by dividing the maximum value
    /// </summary>
    public sealed class Normalize : IAmplitudeConstraint
    {
        EmissionConstraintPtr IAmplitudeConstraint.Ptr() => NativeMethodsGainHolo.AUTDGainHoloConstraintNormalize();
    }

    /// <summary>
    /// Set all amplitudes to the specified value
    /// </summary>
    public sealed class Uniform : IAmplitudeConstraint
    {

        internal readonly EmitIntensity Value;

        public Uniform(byte value)
        {
            Value = new EmitIntensity(value);
        }
        public Uniform(EmitIntensity value)
        {
            Value = value;
        }

        EmissionConstraintPtr IAmplitudeConstraint.Ptr() => NativeMethodsGainHolo.AUTDGainHoloConstraintUniform(Value.Value);
    }

    /// <summary>
    /// Clamp all amplitudes to the specified range
    /// </summary>
    public sealed class Clamp : IAmplitudeConstraint
    {
        internal readonly EmitIntensity Min;
        internal readonly EmitIntensity Max;

        public Clamp(EmitIntensity min, EmitIntensity max)
        {
            Min = min;
            Max = max;
        }

        EmissionConstraintPtr IAmplitudeConstraint.Ptr() => NativeMethodsGainHolo.AUTDGainHoloConstraintClamp(Min.Value, Max.Value);
    }
}
